# Databases and resources 
<!-- â˜… -->

Bioinformatics databases are crucial resources that provide access to a vast array of biological data, ranging from genetic sequences and protein structures to functional annotations and disease-related information. These databases play a pivotal role in various research fields, aiding scientists in data analysis, hypothesis testing, and discovery. Here's an overview of some of the most important bioinformatics databases and how to use them:

**GenBank:** GenBank is a comprehensive database maintained by the National Center for Biotechnology Information (NCBI) that contains DNA sequences submitted by researchers worldwide. It includes sequences from various organisms, along with associated metadata and annotations. Users can search for specific sequences, access genome assemblies, and retrieve information for various genomic regions.

Key Features of GenBank:

Sequence Collection: GenBank contains a vast collection of DNA, RNA, and protein sequences from diverse organisms, including viruses, bacteria, fungi, plants, animals, and more.

Annotations: Sequences in GenBank are accompanied by annotations that provide valuable information about the sequence, such as gene names, protein products, functional regions, and other relevant data.

Genomic Assemblies: GenBank hosts whole genome assemblies for many organisms, enabling researchers to access and analyze complete genomes.

Versioning: Each sequence in GenBank is assigned a version number to track changes and updates. This allows researchers to access historical versions of sequences.

Cross-References: GenBank cross-references other databases and resources, facilitating integration and collaboration among different data sources.

BioProject and BioSample: GenBank integrates with BioProject and BioSample databases, providing additional context for submitted sequences, such as information about the source organism and experimental details.

Search and Retrieval: The GenBank website offers a user-friendly interface for searching and retrieving sequences based on keywords, accession numbers, organisms, and more.




**UniProt:** UniProt is a database that provides detailed information about protein sequences and their functional annotations. It integrates data from various resources, offering a centralized repository for protein-related information. Researchers can search for proteins by name, sequence, or keywords and access data such as protein function, structure, domains, and post-translational modifications.

UniProt (Universal Protein Resource) is a comprehensive and widely used bioinformatics database that provides a centralized repository of protein sequences and functional information. Maintained collaboratively by the European Bioinformatics Institute (EBI), the Swiss Institute of Bioinformatics (SIB), and the Protein Information Resource (PIR), UniProt serves as a valuable resource for researchers, educators, and students in the fields of biology, genetics, and molecular biology.

Key Features of UniProt:

Protein Sequences: UniProt contains a vast collection of protein sequences from a wide range of organisms, including bacteria, archaea, fungi, plants, animals, and viruses.

Functional Annotations: Each protein entry in UniProt is accompanied by extensive functional annotations, providing information about the protein's biological role, structure, domains, post-translational modifications, interactions, subcellular localization, and more.

Cross-References: UniProt integrates information from other databases and resources, offering cross-references to relevant data sources, such as gene databases, literature citations, and 3D protein structures.

Isoforms and Variants: UniProt includes information about protein isoforms, splice variants, and natural variants that affect protein function.

Taxonomic Information: Protein entries are categorized based on taxonomy, enabling users to explore protein diversity across different organisms.

Sequence Alignments: UniProt provides sequence alignments for families of homologous proteins, aiding in understanding protein evolution and relationships.

Disease Annotations: UniProt includes annotations linking proteins to various diseases, such as genetic disorders and cancers, based on experimental evidence.

Literature Citations: Protein entries often include references to scientific literature, allowing users to access the original research articles related to specific proteins.

Keywords and Annotations: UniProt uses controlled vocabularies to assign keywords and annotations to proteins, facilitating standardized and structured data representation.

Components of UniProt:

UniProt is divided into three main components:

UniProtKB: This is the primary component, providing a comprehensive collection of protein sequences and annotations. UniProtKB is further subdivided into three subdatabases:

UniProtKB/Swiss-Prot: Curated protein entries with high-quality annotations.
UniProtKB/TrEMBL: Automatically annotated protein entries that have not yet been curated.
UniRef: This component clusters closely related protein sequences into UniRef clusters, reducing redundancy and simplifying sequence analysis.

UniParc: UniParc contains a unique and non-redundant collection of all publicly available protein sequences, regardless of their level of annotation.

Usage of UniProt:

To use UniProt effectively:

Search: Enter protein names, identifiers, gene names, keywords, or organisms in the search bar.
Browse: Explore protein entries by taxonomy, organism, keywords, and functional annotations.
Retrieve Information: Access protein sequences, functional annotations, cross-references, and literature citations.
Analyze: Utilize the provided data to understand protein function, structure, interactions, and disease associations.
Integration: Integrate UniProt data with other bioinformatics tools and resources for comprehensive analyses.
UniProt is an invaluable resource for researchers studying proteins, their functions, and their roles in various biological processes. Its user-friendly interface, extensive annotations, and cross-referencing capabilities make it a cornerstone tool in the field of molecular biology.

**PubMed:** PubMed is a database maintained by the National Library of Medicine, offering access to an extensive collection of biomedical literature. Researchers can search for articles related to specific topics, diseases, genes, or proteins. PubMed provides abstracts, full-text articles, and links to external resources, aiding literature review and research planning.

**Ensembl:** Ensembl is a genome browser and annotation database that provides genomic sequences, gene annotations, and functional information for a wide range of species. Users can explore gene structures, regulatory elements, genetic variations, and comparative genomics data. Ensembl's browser interface allows for interactive exploration of genomic regions.

**NCBI Gene:** This database offers comprehensive information about genes, including functional annotations, expression data, gene-disease associations, and orthologous relationships. Researchers can search for specific genes, access genetic and protein sequences, and retrieve details about gene function and expression patterns.

The NCBI Gene database, maintained by the National Center for Biotechnology Information (NCBI), is a comprehensive resource that provides information about genes from a wide range of organisms. It serves as a centralized repository for gene-related data, annotations, and functional information. The database plays a crucial role in molecular biology research, aiding scientists in understanding gene functions, interactions, and regulatory mechanisms.

Key Features of the NCBI Gene Database:

Gene Information: The NCBI Gene database contains detailed information about individual genes, including gene names, symbols, aliases, genomic locations, chromosomal coordinates, and orientations.

Gene Function: Each gene entry is accompanied by functional annotations that describe the gene's biological roles, molecular functions, and involvement in cellular processes.

Transcripts and Isoforms: The database provides information about different transcript variants and isoforms associated with a gene. This includes details about alternative splicing, coding regions, and untranslated regions.

Orthologs and Paralogs: NCBI Gene includes information about orthologous and paralogous genes across different species, aiding in understanding gene evolution and relationships.

Homology and Alignment: Gene entries often include sequence alignments, similarity scores, and evolutionary relationships with related genes.

Protein Products: NCBI Gene provides information about the protein products encoded by genes, including functional domains, post-translational modifications, and protein interactions.

Gene Ontology (GO) Annotations: Many gene entries are associated with Gene Ontology terms that categorize gene functions, cellular components, and biological processes.

Expression Data: The database offers data on gene expression patterns in various tissues, developmental stages, and conditions, aiding in understanding gene regulation.

Genomic Context: Gene entries include information about neighboring genes, regulatory elements, and genetic variations in the genomic vicinity.

Disease Associations: Gene entries may include annotations linking genes to specific diseases, based on experimental evidence.

Usage of the NCBI Gene Database:

To use the NCBI Gene database effectively:

Search: Enter gene names, symbols, IDs, or keywords in the search bar.
Browse: Explore gene entries by species, functional annotations, and genomic locations.
Retrieve Information: Access gene details, functional annotations, orthologs, paralogs, and protein products.
Analyze: Utilize the provided data to understand gene functions, expression patterns, and disease associations.
Cross-Reference: NCBI Gene integrates with other NCBI databases, providing links to nucleotide sequences, protein sequences, PubMed articles, and more.
The NCBI Gene database serves as a critical resource for researchers studying individual genes and their roles in various biological processes. Its comprehensive annotations, integration with other NCBI resources, and user-friendly interface make it an essential tool for molecular biologists, geneticists, and researchers across diverse fields.






**dbSNP:** The Single Nucleotide Polymorphism Database (dbSNP) is a repository of genetic variations, including SNPs and other types of variations. Researchers can search for specific variations, retrieve their annotations, and explore data related to population frequencies, diseases, and functional effects.

The dbSNP (Single Nucleotide Polymorphism Database) is a comprehensive bioinformatics resource maintained by the National Center for Biotechnology Information (NCBI). It serves as a repository for genetic variations, including single nucleotide polymorphisms (SNPs), insertions, deletions, and other types of genetic variants. dbSNP plays a crucial role in cataloging and providing access to genetic variations across various species, aiding researchers in understanding the genetic basis of traits, diseases, and population diversity.

Key Features of the dbSNP Database:

Variation Catalog: dbSNP contains a wide variety of genetic variations, including SNPs, short insertions and deletions (indels), microsatellites, and larger structural variants.

Variant Annotations: Each genetic variant is accompanied by annotations, such as genomic coordinates, alleles, frequencies in different populations, and clinical significance (if applicable).

Population Data: dbSNP provides information about the frequency of genetic variants in different human populations, enabling researchers to study population genetics and diversity.

Functional Annotations: Many genetic variants are annotated with potential functional consequences, such as effects on protein coding, splicing, or regulatory elements.

Cross-References: dbSNP cross-references other databases, allowing users to access additional information about genes, proteins, and diseases associated with specific genetic variants.

Genomic Context: The database includes information about the genomic context of variations, such as neighboring genes, regulatory elements, and conserved regions.

ClinVar Integration: dbSNP integrates with the ClinVar database, providing information about the clinical significance of genetic variants and their associations with diseases.

Genome Build Compatibility: dbSNP supports different genome builds, enabling users to map variations to specific versions of the genome.

Usage of the dbSNP Database:

To use the dbSNP database effectively:

Search: Enter SNP IDs, genomic coordinates, gene names, or keywords in the search bar.
Browse: Explore variants by chromosome, population, frequency, or functional annotations.
Retrieve Information: Access variant details, genomic coordinates, allele frequencies, and clinical significance information.
Analyze: Utilize variant annotations for studying disease associations, population genetics, and functional effects.
Integration: Combine dbSNP data with other resources to enhance genetic and genomic analyses.
dbSNP is a crucial resource for researchers investigating the genetic variations that contribute to phenotypic differences, diseases, and population diversity. Its extensive collection of annotated variants, integration with other databases, and user-friendly interface make it an indispensable tool for geneticists, epidemiologists, and researchers in related fields.

**OMIM:** Online Mendelian Inheritance in Man (OMIM) is a database that catalogs genetic mutations and their relationships to inherited diseases. Researchers can search for genes, phenotypes, and diseases to access information about genetic disorders, associated mutations, and relevant literature.

The Online Mendelian Inheritance in Man (OMIM) database is a comprehensive and authoritative resource that catalogues information about genetic disorders, traits, and other phenotypic traits in humans. Maintained by the McKusick-Nathans Institute of Genetic Medicine at Johns Hopkins University, OMIM plays a pivotal role in bridging the gap between genetic research and clinical medicine, providing valuable insights into the genetic basis of various human conditions.

Key Features of the OMIM Database:

Genetic Disorders: OMIM contains detailed information about a wide range of genetic disorders, both rare and common, providing insights into the genetic mutations, inheritance patterns, clinical features, and molecular mechanisms underlying these conditions.

Phenotypic Traits: In addition to genetic disorders, OMIM also catalogues information about various phenotypic traits, such as physical characteristics, susceptibility to diseases, and responses to treatments.

Gene-Centric Information: OMIM provides gene-centric entries that include information about specific genes associated with genetic disorders. These entries detail gene functions, mutations, protein products, and any known associations with diseases or traits.

Clinical Descriptions: Each entry includes detailed clinical descriptions of the disorders or traits, including symptoms, diagnostic criteria, and information about disease progression.

Genetic Inheritance Patterns: OMIM categorizes disorders based on their inheritance patterns, such as autosomal dominant, autosomal recessive, X-linked, and mitochondrial.

Gene and Locus References: OMIM cross-references genes and loci to other databases and resources, facilitating integration with genomic and functional data.

Links to Literature: OMIM entries include references to scientific literature, providing access to primary research articles and reviews related to specific disorders and genes.

Molecular Mechanisms: Many OMIM entries describe the molecular mechanisms underlying genetic disorders, providing insights into how mutations affect biological processes.

Usage of the OMIM Database:

To use the OMIM database effectively:

Search: Enter disease names, gene names, phenotypic keywords, or gene identifiers in the search bar.
Browse: Explore entries by disease categories, gene names, inheritance patterns, and other criteria.
Retrieve Information: Access detailed information about genetic disorders, phenotypic traits, genes, clinical descriptions, and molecular mechanisms.
Analyze: Utilize the provided data to understand the genetic basis of diseases, inheritance patterns, and potential therapeutic targets.
Integration: Combine OMIM data with other genetic and genomic resources to enhance research and clinical applications.
OMIM is widely used by geneticists, clinicians, researchers, and healthcare professionals to gain insights into the genetic underpinnings of diseases, guide clinical diagnoses, and inform therapeutic strategies. Its rich content, authoritative sources, and user-friendly interface make it an essential tool for anyone involved in genetics and genomics research.


**STRING:** The STRING database provides information about protein-protein interactions and functional associations. Users can search for a protein of interest to explore its interactions, view network maps, and analyze the potential functions of proteins within a network context.

To use these databases effectively:

Keyword Search: Start by entering relevant keywords, gene names, protein identifiers, or disease names in the search bar.
Filters: Many databases offer filters to refine search results based on criteria such as species, data type, and publication date.
Advanced Search: Use advanced search options to specify specific fields or search criteria to narrow down results.
Data Retrieval: Once you find relevant information, you can often download sequences, annotations, and other data for further analysis.
Citations: When using data from these databases in your research, make sure to provide proper citations to acknowledge the original sources.
These databases are continuously updated and expanded, so it's essential to explore their documentation and tutorials to stay informed about the latest features and search strategies.

The STRING database (Search Tool for the Retrieval of Interacting Genes/Proteins) is a valuable bioinformatics resource that provides information about protein-protein interactions (PPIs) and functional associations. STRING helps researchers explore the relationships between proteins, gain insights into biological processes, and understand the complex networks that underlie various cellular functions. Maintained by a collaborative team of researchers, STRING is widely used in various fields, including molecular biology, systems biology, and drug discovery.

Key Features of the STRING Database:

Protein Interaction Data: STRING aggregates and integrates experimental and computational data on protein interactions from diverse sources, including high-throughput experiments, curated databases, and literature mining.

Functional Associations: Beyond direct physical interactions, STRING provides information about functional associations between proteins, such as co-expression, shared domains, and similar annotations.

Confidence Scores: STRING assigns confidence scores to interactions and associations based on the evidence supporting them. These scores help users assess the reliability of the reported interactions.

Network Visualization: STRING generates interactive graphical representations of protein networks, allowing users to visualize and explore protein interactions and functional relationships in a visual format.

Species Coverage: STRING supports a wide range of organisms, from model organisms to less-studied species, enabling researchers to investigate protein interactions in different biological contexts.

Enrichment Analysis: STRING offers enrichment analysis tools that help users identify overrepresented functional categories, pathways, and Gene Ontology terms within a set of proteins of interest.

Prediction Methods: STRING includes computational prediction methods that estimate potential interactions based on sequence similarities, domain architectures, and other features.

Cross-References: STRING provides cross-references to external resources, allowing users to access additional information about interacting proteins, pathways, and diseases.

Usage of the STRING Database:

To use the STRING database effectively:

Search: Enter protein names, identifiers, or keywords to find interactions and associations.
Network Visualization: Explore and visualize protein interaction networks, adjusting confidence thresholds and interaction types.
Retrieve Information: Access interaction scores, functional annotations, pathway information, and external references.
Analyze: Utilize the provided data to understand protein functions, regulatory mechanisms, and disease associations.
Enrichment Analysis: Perform enrichment analysis to identify functional themes within a set of proteins.
Integration: Combine STRING data with other omics data to gain a holistic understanding of cellular processes.
STRING is a versatile tool for researchers investigating protein interactions, cellular pathways, and systems-level biology. Its user-friendly interface, comprehensive data sources, and advanced analysis features make it an essential resource for studying the complex molecular networks that govern biological functions.

## Sequence databases {-}

## Genome browsers {-}

The UCSC Genome Browser is a widely used and comprehensive online tool for visualizing and exploring genomic information from a wide range of organisms. Developed and maintained by the University of California, Santa Cruz (UCSC), this browser offers a user-friendly interface to navigate and analyze genomes, gene structures, regulatory elements, genetic variations, and other genomic features. Here's an overview of the UCSC Genome Browser and its key features:

Features of the UCSC Genome Browser:

Genome Selection: The UCSC Genome Browser supports a vast collection of genomes from various species, including humans, model organisms, and microbes. Users can select the genome of interest from a dropdown menu.

Genome Views: The browser provides different views of the genome, including the standard linear view, a circular view (for some genomes), and a conservation view that highlights evolutionarily conserved regions.

Navigation and Zoom: Users can easily navigate through genomic regions using navigation arrows and zoom controls. This allows detailed examination of specific regions.

Annotation Tracks: One of the most powerful features of the UCSC Genome Browser is the ability to overlay various types of annotation tracks onto the genome view. These tracks include gene annotations, regulatory elements, genetic variations, and more. Users can customize which tracks are displayed and adjust their visibility and order.

Gene Annotations: The browser displays gene structures, including exons, introns, coding sequences, and untranslated regions. It provides information about gene names, functions, and alternative splicing patterns.

Regulatory Elements: Regulatory elements such as promoters, enhancers, and transcription factor binding sites can be visualized, aiding in understanding gene regulation.

Genetic Variations: Genetic variations, including single nucleotide polymorphisms (SNPs), insertions, and deletions, can be displayed. Users can examine variations' positions, alleles, frequencies, and potential functional consequences.

Custom Tracks: Users can upload their own data, such as experimental results or custom annotations, to visualize alongside the provided tracks.

Comparative Genomics: The browser offers tools to compare genomic sequences and annotations across multiple species. This aids in identifying conserved regions and evolutionary changes.

Search and Highlight: Users can search for specific genes, regions, sequences, or annotations and highlight them in the genome view for detailed examination.

Links to External Resources: The UCSC Genome Browser provides links to external databases and resources, enabling seamless access to additional information.

Usage:

To use the UCSC Genome Browser:

Access the Browser: Go to the UCSC Genome Browser website (genome.ucsc.edu).

Choose a Genome: Select the genome of interest from the dropdown menu.

Navigate and Explore: Navigate to specific genomic regions using the search bar, navigation arrows, and zoom controls. Customize the displayed tracks to focus on relevant features.

Analyze and Download: Analyze gene structures, regulatory elements, and genetic variations in the context of the genome. Download data and images for use in research or presentations.

The UCSC Genome Browser serves as an invaluable resource for researchers, educators, and students seeking to understand and analyze genomic information across various organisms. Its intuitive interface, extensive collection of annotation tracks, and powerful visualization capabilities make it a cornerstone tool in the field of genomics.


## list of databases {-}

GenBank
UniProt
PubMed
Ensembl
NCBI Gene
STRING
dbSNP
OMIM
PDB (Protein Data Bank)
KEGG (Kyoto Encyclopedia of Genes and Genomes)
Reactome
GO (Gene Ontology)
HPRD (Human Protein Reference Database)
COSMIC (Catalogue Of Somatic Mutations In Cancer)
ClinVar
TCGA (The Cancer Genome Atlas)
Pfam
Rfam
InterPro
GTEx (Genotype-Tissue Expression Project)
HGNC (HUGO Gene Nomenclature Committee)
FlyBase
WormBase
TAIR (The Arabidopsis Information Resource)
Mouse Genome Informatics (MGI)
RGD (Rat Genome Database)
PharmGKB
DrugBank
PubChem
dbGaP (Database of Genotypes and Phenotypes)
dbCAN (Carbohydrate-Active enZYmes Database)
CATH (Class, Architecture, Topology, Homology)
MEROPS (Peptidase Database)
BioGRID
IntAct
IUPHAR/BPS Guide to Pharmacology
GenAtlas
miRBase
lncRNAdb
DGV (Database of Genomic Variants)
UCSC Genome Browser
ExPASy (Expert Protein Analysis System)
Swiss-Model
Pfam
SUPERFAMILY
EMBL-EBI InterProScan
WikiPathways
ChEMBL
BioCyc


1. [GenBank](https://www.ncbi.nlm.nih.gov/genbank/)
2. [UniProt](https://www.uniprot.org/)
3. [PubMed](https://pubmed.ncbi.nlm.nih.gov/)
4. [Ensembl](https://www.ensembl.org/)
5. [NCBI Gene](https://www.ncbi.nlm.nih.gov/gene)
6. [STRING](https://string-db.org/)
7. [dbSNP](https://www.ncbi.nlm.nih.gov/snp/)
8. [OMIM](https://www.omim.org/)
9. [PDB (Protein Data Bank)](https://www.rcsb.org/)
10. [KEGG (Kyoto Encyclopedia of Genes and Genomes)](https://www.genome.jp/kegg/)
11. [Reactome](https://reactome.org/)
12. [GO (Gene Ontology)](http://geneontology.org/)
13. [HPRD (Human Protein Reference Database)](http://www.hprd.org/)
14. [COSMIC (Catalogue Of Somatic Mutations In Cancer)](https://cancer.sanger.ac.uk/cosmic)
15. [ClinVar](https://www.ncbi.nlm.nih.gov/clinvar/)
16. [TCGA (The Cancer Genome Atlas)](https://www.cancer.gov/about-nci/organization/ccg/research/structural-genomics/tcga)
17. [Pfam](https://pfam.xfam.org/)
18. [Rfam](https://rfam.xfam.org/)
19. [InterPro](https://www.ebi.ac.uk/interpro/)
20. [GTEx (Genotype-Tissue Expression Project)](https://gtexportal.org/home/)
21. [HGNC (HUGO Gene Nomenclature Committee)](https://www.genenames.org/)
22. [FlyBase](https://flybase.org/)
23. [WormBase](https://www.wormbase.org/)
24. [TAIR (The Arabidopsis Information Resource)](https://www.arabidopsis.org/)
25. [Mouse Genome Informatics (MGI)](http://www.informatics.jax.org/)
26. [RGD (Rat Genome Database)](https://rgd.mcw.edu/)
27. [PharmGKB](https://www.pharmgkb.org/)
28. [DrugBank](https://go.drugbank.com/)
29. [PubChem](https://pubchem.ncbi.nlm.nih.gov/)
30. [dbGaP (Database of Genotypes and Phenotypes)](https://www.ncbi.nlm.nih.gov/gap)
31. [dbCAN (Carbohydrate-Active enZYmes Database)](http://bcb.unl.edu/dbCAN/)
32. [CATH (Class, Architecture, Topology, Homology)](http://www.cathdb.info/)
33. [MEROPS (Peptidase Database)](https://www.ebi.ac.uk/merops/)
34. [BioGRID](https://thebiogrid.org/)
35. [IntAct](https://www.ebi.ac.uk/intact/)
36. [IUPHAR/BPS Guide to Pharmacology](https://www.guidetopharmacology.org/)
37. [GenAtlas](https://genatlas.medecine.univ-paris5.fr/)
38. [miRBase](http://www.mirbase.org/)
39. [lncRNAdb](http://www.lncrnadb.org/)
40. [DGV (Database of Genomic Variants)](http://dgv.tcag.ca/)
41. [UCSC Genome Browser](https://genome.ucsc.edu/)
42. [ExPASy (Expert Protein Analysis System)](https://web.expasy.org/)
43. [Swiss-Model](https://swissmodel.expasy.org/)
44. [Pfam](https://pfam.xfam.org/)
45. [SUPERFAMILY](http://supfam.cs.bris.ac.uk/)
46. [EMBL-EBI InterProScan](https://www.ebi.ac.uk/interpro/interproscan.html)
47. [WikiPathways](https://www.wikipathways.org/)
48. [ChEMBL](https://www.ebi.ac.uk/chembl/)
49. [BioCyc](https://biocyc.org/)
50. [dbVar](https://www.ncbi.nlm.nih.gov/dbvar/)




## Knowledege data bases {-}

## Do it yourself:  {-}


